src/targets/__tests__/maven.test.ts

```diff
@@ -0,0 +1,212 @@
+          [secret]: DEFAULT_OPTION_VALUE,
+        })
+      )
+    );
+  });
+
+  test('without options', () =>
+    expect(createMavenTarget).toThrowErrorMatchingInlineSnapshot(
+      `"Required value(s) OSSRH_USERNAME not found in configuration files or the environment. See the documentation for more details."`
+    ));
+});
+
+describe('publish', () => {
+  const tmpDirName = 'tmpDir';
+
+  beforeAll(() => setTargetSecretsInEnv());
+
+  afterAll(() => removeTargetSecretsFromEnv());
+
+  beforeEach(() => jest.resetAllMocks());
+
+  test('main flow', async () => {
+    const callOrder: string[] = [];
+    const mvnTarget = createMavenTarget();
+    const gradlePropsMock = jest.fn(
+      async () => void callOrder.push('gradleProps')
+    );
+    mvnTarget.createUserGradlePropsFile = gradlePropsMock;
+    const uploadMock = jest.fn(async () => void callOrder.push('upload'));
+    mvnTarget.upload = uploadMock;
+    (retrySpawnProcess as jest.MockedFunction<
+      typeof retrySpawnProcess
+    >).mockImplementationOnce(
+      async () => void callOrder.push('closeAndRelease')
+    );
+
+    const version = '1.0.0';
+    const revision = 'r3v1s10n';
+
+    await mvnTarget.publish(version, revision);
+    expect(gradlePropsMock).toHaveBeenCalledTimes(1);
+    expect(uploadMock).toHaveBeenCalledTimes(1);
+    expect(uploadMock).toHaveBeenLastCalledWith(revision);
+    expect(retrySpawnProcess).toHaveBeenCalledTimes(1);
+    expect(retrySpawnProcess).toHaveBeenCalledWith(DEFAULT_OPTION_VALUE, [
+      'closeAndReleaseRepository',
+    ]);
+    expect(callOrder).toStrictEqual([
+      'gradleProps',
+      'upload',
+      'closeAndRelease',
+    ]);
+  });
+
+  test('upload', async () => {
+    // simple mock to always use the same temporary directory,
+    // instead of creating a new one
+    (withTempDir as jest.MockedFunction<typeof withTempDir>).mockImplementation(
+      async cb => {
+        return await cb(tmpDirName);
+      }
+    );
+
+    const mvnTarget = createMavenTarget();
+    mvnTarget.getArtifactsForRevision = jest
+      .fn()
+      .mockResolvedValueOnce([{ filename: 'mockArtifact.zip' }]);
+    mvnTarget.artifactProvider.downloadArtifact = jest
+      .fn()
+      .mockResolvedValueOnce('artifact/download/path');
+
+    await mvnTarget.upload('r3v1s10n');
+    expect(retrySpawnProcess).toBeCalledTimes(1);
+    const callArgs = (retrySpawnProcess as jest.MockedFunction<
+      typeof retrySpawnProcess
+    >).mock.calls[0];
+
+    expect(callArgs).toHaveLength(2);
+    expect(callArgs[0]).toBeDefined();
+    expect(callArgs[0]).toEqual(DEFAULT_OPTION_VALUE);
+
+    const cmdArgs = callArgs[1];
+    expect(cmdArgs).toBeDefined();
+    expect(cmdArgs).toHaveLength(10);
+    // @ts-ignore `cmdArgs[*]` possibly undefined
+    expect(cmdArgs[0]).toEqual('gpg:sign-and-deploy-file');
+    // @ts-ignore `cmdArgs[*]` possibly undefined
+    expect(cmdArgs[1]).toMatch(new RegExp(`-Dfile=${tmpDirName}.*`));
```

Comment:
One worry: the order of these arguments does not matter in real life does it? But then a snapshot test would have the same issue.

Better to be more explicit about we expect _something_ after the dirname.
```suggestion
    expect(cmdArgs[1]).toMatch(new RegExp(`-Dfile=${tmpDirName}.+`));
```

Reply:
Yes, it shouldn't matter. Since testing regex-s like this is the only way I managed to get it tested, I thought it wasn't worth it to do an algo for it.